core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java

```diff
@@ -0,0 +1,239 @@
+     * @param seqType ending option either a specific/absolute value, or last message as the end
+     */
+    private MessageNumberSet(final long start, final long end, @Nonnull final SequenceType seqType) {
+        this.start = start;
+        this.end = end;
+        this.seqType = seqType;
+    }
+
+    /**
+     * Converts an array of integers into an array of MessageNumberSet. This is a helper method for callers with int array.
+     *
+     * @param msgs array of primitive integer data type message number (could be message sequence or UID)
+     * @return MessageNumberSet array
+     */
+    public static MessageNumberSet[] createMessageNumberSets(@Nonnull final int[] msgs) {
+        List<MessageNumberSet> v = new ArrayList<MessageNumberSet>();
+        int i, j;
+        for (i = 0; i < msgs.length; i++) {
+            int start = msgs[i];
+
+            // Look for contiguous elements
+            for (j = i + 1; j < msgs.length; j++) {
+                if (msgs[j] != msgs[j - 1] + 1) {
+                    break;
+                }
+            }
+            int end = msgs[j - 1];
+            v.add(new MessageNumberSet(start, end));
+            i = j - 1; // i gets incremented @ top of the loop
+        }
+        return v.toArray(new MessageNumberSet[v.size()]);
+    }
+
+    /**
+     * Converts an array of long into array of MessageNumberSet. This is a helper method for callers with long array.
+     *
+     * @param msgs array of long data type message number (could be message sequence or UID)
+     * @return the string generated that conforms to RFC3501 Message Number syntax
+     */
+    public static MessageNumberSet[] createMessageNumberSets(@Nonnull final long[] msgs) {
+        List<MessageNumberSet> v = new ArrayList<MessageNumberSet>();
+        int i, j;
+        for (i = 0; i < msgs.length; i++) {
+            long start = msgs[i];
+
+            // Look for contiguous elements
+            for (j = i + 1; j < msgs.length; j++) {
+                if (msgs[j] != msgs[j - 1] + 1) {
+                    break;
+                }
+            }
+            long end = msgs[j - 1];
+            v.add(new MessageNumberSet(start, end));
+            i = j - 1; // i gets incremented @ top of the loop
+        }
+        return v.toArray(new MessageNumberSet[v.size()]);
+    }
+
+    /**
+     * Converts an array of MessageNumberSet into an IMAP RFC3501 sequence-set syntax.
+     *
+     * @param msgsets array of MessageNumberSet
+     * @return the string generated that conforms to IMAP RFC3501 sequence-set syntax
+     */
+    public static String buildString(@Nullable final MessageNumberSet[] msgsets) {
+        if (msgsets == null || msgsets.length == 0) {
+            return null;
+        }
+
+        int i = 0; // msgset index
+        final StringBuilder s = new StringBuilder();
+        int size = msgsets.length;
+        long start, end;
+
+        for (;;) {
```

Comment:
Infinite loops are not good. I would have a terminating condition.

Reply:
Done